### PR TITLE
The World's Smallest Concourse CI Server

### PR DESCRIPTION
**Pull request by cunnie**
A blog post describing how to deploy a small Concourse server to Amazon AWS.

@pivotal/pivotal-engineering-directors , if you have any suggestions, let me know; however, I am aware of the many demands on your time, so I don't expect you to review it; instead, I plan to review the post with @robdimsdale of the Concourse team.

Known bugs:
* the screenshots are of different sizes, an artifact of taking screenshots both from a retina display & non-retina display
* The Concourse server doesn't work because it doesn't have any workers. That's going to be a follow-on blog post. This blog-post is already > 1,500 words and I didn't want to make it any longer.

### CODE REVIEW

File: content/post/worlds-smallest-concourse-server.md [hunk 1]
```diff
@@ -0,0 +1,287 @@
+---
+author: cunnie
+categories:
+- BOSH
```
Comment (robdimsdale) [inline content/post/worlds-smallest-concourse-server.md]: Is there a concourse category? If not we should probably create one.

File: content/post/worlds-smallest-concourse-server.md [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+
+* click **Save**
+
+### 0.2 Obtain SSL Certificates
+
+We decide to use HTTPS to communicate with our Concourse server, for we will
+need to authenticate against the webserver when we configure our CI
+(when we transmit our credentials over the Internet we want them to be encrypted).
+
+We purchase valid SSL certificates for our server. Using a self-signed certificate is also an option.
```
Comment (robdimsdale) [inline content/post/worlds-smallest-concourse-server.md]: As you are probably aware, [let's encrypt](https://letsencrypt.org/) offers free SSL certificates. They're not available just yet, but should be usable in the next month or so.

File: content/post/worlds-smallest-concourse-server.md [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+* download the redacted (passwords & keys removed) BOSH [manifest](https://github.com/blabbertabber/blabbertabber/blob/develop/ci/concourse-aws.yml)
+* open it in an editor
+* search for every occurrence of *FIXME*
+* update the field described by the *FIXME*, e.g. update the IP address,
+  set the password, set the Subnet ID, etc....
+
+For those interested, our sample Concourse manifest was derived from
+Concourse's official sample *bosh-init* [manifest](https://github.com/concourse/concourse/blob/master/manifests/bosh-init-aws.yml) and modified as follows:
+
+* added an nginx release to avoid the cost of an ELB ($219.14/year <sup>[[ELB-pricing]](#ELB-pricing)</sup> ).
```
Comment (robdimsdale) [inline content/post/worlds-smallest-concourse-server.md]: Is the point of adding nginx to provide SSL termination? It's obviously not for load balancing with two ATCs

File: content/post/worlds-smallest-concourse-server.md [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+for reviewing the manuscript for accuracy and completeness.
+
+## Footnotes
+
+<a name="snowflakes"><sup>[snowflakes]</sup></a>
+Even the most innocuous changes to a CI server can be fraught with anxiety: two years
+ago when we were migrating one of our development team's Jenkins CI server VM
+from one datastore to another (a very low-risk operation),
+we needed to have several meetings with the Team's
+project manager and the anchor before they were willing to allow us to proceed
```
Comment (robdimsdale) [inline content/post/worlds-smallest-concourse-server.md]: project or product manager?

File: content/post/worlds-smallest-concourse-server.md [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+stores its state on a [persistent disk](https://bosh.io/docs/persistent-disks.html),
+so that re-deploying the CI server (e.g. new OS, new Concourse) won't cause the loss
+of the pipeline configuration or build history.
+
+## Appendix A. Concourse Yearly Costs: $80.34
+
+The yearly cost of running a Concourse server is $80.34. Note that this
+does not include the cost of the worker.
+Had we chosen to implement the recommended m3.large EC2 instance for a worker, it
+would have increased our yearly cost by $713.54 <sup>[[m3.large]](#m3.large)</sup> .
```
Comment (robdimsdale) [inline content/post/worlds-smallest-concourse-server.md]: I assume the follow-on blog post will determine a reasonable instance type for the worker, such that it doesn't increase the total cost by ~1400%

**Comment by robdimsdale** [comment]
The only other thing might be to actually set up a worker to provide a better out-of-box experience. However, it sounds like this isn't your target audience, so I think it's fine either way.

**Comment by robdimsdale** [comment]
Otherwise, everything looks great from a technical perspective.

File: content/post/worlds-smallest-concourse-server.md (hunk 1 above)
Comment (cunnie) [inline content/post/worlds-smallest-concourse-server.md]: Done.

File: content/post/worlds-smallest-concourse-server.md (hunk 5 above)
Comment (cunnie) [inline content/post/worlds-smallest-concourse-server.md]: Correct. The next blog post will be instructions to installing local VMware Fusion-based workers (there's a reason for this: nested hardware virtualization).

**Comment by cunnie** [comment]
@tsaleh Could you please merge my pull request (approve the post)? I've incorporated all of @robdimsdale 's most-excellent suggestions and squashed my commits so there's only one. Thanks, you rock!

**Comment by robdimsdale** [comment]
I'm happy with all the changes.

For your reference, Brian, the MEGA team is actively working on tooling to improve the user-experience of creating the AWS stack for concourse, but we're not ready for public consumption of that tooling just yet.

When that tooling is sufficiently mature, you may wish to update this article to point readers at that instead of making their own.

**Comment by tsaleh** [comment]
Thanks @cunnie - I'll review the content on my flight, and hopefully publish it after.  One quick change, though:  can we include those images in the commit, so they're served directly, instead of relying on an external image hosting service?

Also, maybe include a bit addressing @robdimsdale's latest comment - that we're working to make this much easier, soon.  @robdimsdale In the interest of ship-early-ship-often, what's the cause for the reticence to release what you have now?  Security issues?

**Comment by cunnie** [comment]
@tsaleh :

1. images are included in the latest squashed commit; they are no longer being served from an external image hosting service.
2. I have incorporated both yours and Rob's comments about the upcoming AWS Concourse tooling in a footnote (referenced from where the user is directed to the bosh.io instructions to create the infrastructure). When the tooling is ready I'll try to incorporate it in the blog post.

Let me know if there's anything else you'd like to see. Thanks again,

—Brian

**Comment by robdimsdale** [comment]
@tsaleh the MEGA tooling does not currently create a suitable stack for a bosh-init'd concourse, and I'm not sure where that might fit into the roadmap. Currently the tooling supports creating the infrastructure for a bosh-init'd director and concourse deployed to that, as well as infrastructure for CF.

The tooling is intended to provided a good out-of-box experience, but we also want it to be easily extensible, such that modifications like opening port 80/443 would fit into the workflow, and not have to be done out of band. This is something we're figuring out how to incorporate, so I'll pass this use-case on to the PM.

Although I haven't tested it, it looks like this blog post should work with the latest version of concourse and fly (`0.66.1`) and so it should be fine from a technical perspective.

**Comment by cunnie** [comment]
To confirm, the blog post is targeted at version `0.66.1` (the most recent Concourse release):

* the screenshots were from a server that was deployed with version `0.66.1`
* the included sample manifest uses release `0.66.1`, including the SHA & URI
* the fly command used was version `0.66.1`

After talking with @robdimsdale, I may include (at a later date) a cloudformation template that could be used to bypass the tedious infrastructure-creation step.

When you're happy with it, @tsaleh , please publish. I am happy with it.

**Comment by tsaleh** [comment]
@cunnie, this was a marvelous post, and well worth the $5!

**Comment by robdimsdale** [comment]
@tsaleh do reviewers also get some dollar?

**Comment by tsaleh** [comment]
Good try, @robdimsdale ;)  But there are four more slots for the coveted FIVE DOLLAR prize!

**Pull request merged**